File: test-suite-kotlin/src/test/kotlin/io/micronaut/picocli/docs/quickstart/GitStarCommand.kt
package io.micronaut.picocli.docs.quickstart

// tag::imports[]
import io.micronaut.configuration.picocli.PicocliRunner
import io.micronaut.http.HttpRequest
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.annotation.Client
import jakarta.inject.Inject
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import picocli.CommandLine.Parameters
import kotlin.system.exitProcess
// end::imports[]

// tag::class[]
@Command(
    name = "git-star",
    header = [
        "@|green       _ _      _             |@",  // <1>
        "@|green  __ _(_) |_ __| |_ __ _ _ _  |@",
        "@|green / _` | |  _(_-<  _/ _` | '_| |@",
        "@|green \\__, |_|\\__/__/\\__\\__,_|_|@",
        "@|green |___/                        |@"],
    description = ["Shows GitHub stars for a project"],
    mixinStandardHelpOptions = true,
    version = ["git-star 0.1"] // <2>
)
class GitStarCommand : Runnable {

    @Inject
    @field:Client("https://api.github.com/")
    lateinit var client: HttpClient // <3>

    @Option(names = ["-v", "--verbose"], description = ["Shows some project details"])
    var verbose = false

    @Parameters(  // <4>
        description = ["One or more GitHub slugs (comma separated) to show stargazers for. Default: \${DEFAULT-VALUE}"],
        split = ",",
        paramLabel = "<owner/repo>"
    )
    var githubSlugs: List<String> = mutableListOf("micronaut-projects/micronaut-core", "remkop/picocli")

    override fun run() { // <5>
        val blockingClient = client.toBlocking()
        githubSlugs.forEach { slug ->
            val httpRequest = HttpRequest.GET<Any>("repos/$slug")
                .header("User-Agent", "remkop-picocli")
            val m = blockingClient.retrieve(httpRequest, Map::class.java)
            println("$slug has ${m["watchers"]} stars")

            if (verbose) {
                println("""Description: ${m["description"]}
                          |License: ${(m["license"] as Map<*,*>)["name"]}
                          |Forks: ${m["forks"]}
                          |Open issues: ${m["open_issues"]}
                          |""".trimMargin())
            }
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val exitCode = PicocliRunner.execute(GitStarCommand::class.java, *args)
            exitProcess(exitCode)
        }
    }
}
// end::class[]

File: test-suite-python/src/test/python/micronaut/picocli/docs/quickstart/GitStarCommand.py
# tag::imports[]
from typing import Annotated

from jakarta.inject import Inject
from java.lang import Runnable, System
from micronaut.configuration.picocli import PicocliRunner
from micronaut.http import HttpRequest
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from picocli.CommandLine import Command, Option, Parameters
# end::imports[]


# tag::class[]
@Command(name="git-star", header=[
    "@|green       _ _      _             |@",  # <1>
    "@|green  __ _(_) |_ __| |_ __ _ _ _  |@",
    "@|green / _` | |  _(_-<  _/ _` | '_| |@",
    "@|green \\__, |_|\\__/__/\\__\\__,_|_|@",
    "@|green |___/                        |@"],
    description="Shows GitHub stars for a project",
    mixinStandardHelpOptions=True,
    version="git-star 0.1")  # <2>
class GitStarCommand(Runnable):

    client: Annotated[HttpClient, Client("https://api.github.com"), Inject]  # <3>

    verbose: Annotated[bool, Option(names=["-v", "--verbose"], description="Shows some project details")] = False

    github_slugs: Annotated[list[str], Parameters(  # <4>
        description=[
            "One or more GitHub slugs (comma separated) to show stargazers for. Default: ${DEFAULT-VALUE}"
        ],
        split=",",
        paramLabel="<owner/repo>"
    )] = ["micronaut-projects/micronaut-core", "remkop/picocli"]

    def run(self) -> None:  # <5>
        blocking_client = self.client.toBlocking()
        for slug in self.github_slugs:
            http_request = HttpRequest.GET("/repos/" + slug) \
                .header("User-Agent", "remkop-picocli")
            m = blocking_client.retrieve(http_request, dict)
            print(f"{slug} has {m['watchers']} stars")

            if self.verbose:
                print(f"Description: {m['description']}\n"
                      f"License: {m['license']['name']}\n"
                      f"Forks: {m['forks']}\n"
                      f"Open issues: {m['open_issues']}\n")

    @staticmethod
    def main(args: list[str]) -> None:
        exit_code = PicocliRunner.execute(GitStarCommand, args)
        System.exit(exit_code)
# end::class[]
